### PR TITLE
Part2

### PR DESCRIPTION
**Pull request by cristianvogel**
:)

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -61,124 +69,411 @@ New/This/Pointers/References conclusion
...
-    float add (float, float);
-    float subtract ( float, float);
-    float multiply ( float, float);
-    float divide (float, float);
+    float* pointerToFloatValue = nullptr;
+
+    // default constructor assigns heap primitive if no value sent in
+    FloatType() {
+        pointerToFloatValue = new float();
+     }
```
Comment (matkatmusic) [inline main.cpp]: - [x] remove default constructors

File: main.cpp [hunk 2]
```diff
@@ -61,124 +69,411 @@ New/This/Pointers/References conclusion
...
-    double subtract ( double, double);
-    double multiply ( double, double);
-    double divide (double, double);
+    double* pointerToDoubleValue = nullptr;
+
+    // default constructor assigns heap primitive if no value sent in
+    DoubleType() {
+        pointerToDoubleValue = new double();
+    } 
+
```
Comment (matkatmusic) [inline main.cpp]: - [x] remove default constructors

File: main.cpp [hunk 3]
```diff
@@ -61,124 +69,411 @@ New/This/Pointers/References conclusion
...
 
-double DoubleType::divide (double x, double y) 
+DoubleType& DoubleType::divide ( const double y) 
 { 
-    if (abs(y) > 0) return (x/y);    
+    if (abs(y) > 0) *pointerToDoubleValue /= y;    
     
     std::cout << "Divide-by-zero warning  " << std::endl;
-    return (x/y);      
+    *pointerToDoubleValue /= 0;
```
Comment (matkatmusic) [inline main.cpp]: you need an `else` here.  you're always dividing by 0. 

- [x] add `else` block, remove division by zero.  floating point division by zero is allowed.

File: main.cpp [hunk 4]
```diff
@@ -61,124 +69,411 @@ New/This/Pointers/References conclusion
...
-    int add (int, int);
-    int subtract ( int, int);
-    int multiply ( int, int);
-    int divide (int, int);
+    int* pointerToIntValue = nullptr;
+
+     // default constructor assigns heap primitive if no value sent in
+    IntType() {
+        pointerToIntValue = new int();
+    }
```
Comment (matkatmusic) [inline main.cpp]: - [x] remove default constructors

File: main.cpp [hunk 5]
```diff
@@ -61,124 +69,411 @@ New/This/Pointers/References conclusion
...
+IntType& IntType::divide ( const int y ) 
 { 
-    return (x*y); 
+    if (y != 0 ) 
+    {
+        *pointerToIntValue /=y; 
+        return *this;
+    } 
+    std::cout << "Cannot Divide Int by Zero! " << std::endl;
+    *pointerToIntValue = 0;
```
Comment (matkatmusic) [inline main.cpp]: Why are you changing the value?   You should just warn the user.

- [x] remove assignment of 0

File: main.cpp [hunk 6]
```diff
@@ -61,124 +69,411 @@ New/This/Pointers/References conclusion
...
-float FloatType::divide (float x, float y) 
+FloatType& FloatType::divide ( const float y ) 
 { 
-    if (abs(y) > 0) return (x/y);    
+    if (abs(y) > 0) *pointerToFloatValue /= y;    
     
     std::cout << "Divide-by-zero warning  " << std::endl;
-    return (x/y);      
+    *pointerToFloatValue /= 0;
+    return *this;
```
Comment (matkatmusic) [inline main.cpp]: you need an `else` here.  you're always dividing by 0. 

- [x] add `else` block, remove division by zero.  floating point division by zero is allowed.

File: main.cpp [hunk 7]
```diff
@@ -61,124 +69,411 @@ New/This/Pointers/References conclusion
...
+{ 
+    if (*dt.pointerToDoubleValue > 0) 
+    {
+        *pointerToIntValue /= *dt.pointerToDoubleValue; 
+        return *this;
+    }
     std::cout << "Cannot Divide Int by Zero! " << std::endl;
-    return 0;
+    *pointerToIntValue = 0;
+    return *this;
```
Comment (matkatmusic) [inline main.cpp]: All of these should just return the result of calling the function that takes the primitive.  
i.e. `return multiply( *dt.pointerToDoubleValue );`

- [x] D.R.Y. refactor

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp [hunk 8]
```diff
@@ -61,124 +69,342 @@ New/This/Pointers/References conclusion
...
+FloatType& FloatType::subtract( const DoubleType& dt ) 
+{ 
+    return subtract( *dt.pointerToDoubleValue ); 
+}
+
+FloatType& FloatType::multiply( const DoubleType& dt) 
+{ 
+    *pointerToFloatValue *= *dt.pointerToDoubleValue;
+    return multiply( *dt.pointerToDoubleValue ); 
+}
```
Comment (matkatmusic) [inline main.cpp]: You're multiplying twice. 

- [x] fixme

File: main.cpp [hunk 9]
```diff
@@ -61,124 +69,342 @@ New/This/Pointers/References conclusion
...
+FloatType& FloatType::add( const IntType& it )
+{
+    return add( *it.pointerToIntValue ) ;
+}
+
+FloatType& FloatType::subtract( const IntType& it ) 
+{ 
+    *pointerToFloatValue -= *it.pointerToIntValue;
+    return subtract( *it.pointerToIntValue ); 
+}
```
Comment (matkatmusic) [inline main.cpp]: you're subtracting twice.

- [x] FIXME

File: main.cpp [hunk 10]
```diff
@@ -61,124 +69,342 @@ New/This/Pointers/References conclusion
...
+
+FloatType& FloatType::multiply( const IntType& it) 
+{ 
+    *pointerToFloatValue *= *it.pointerToIntValue;
+    return multiply( *it.pointerToIntValue )  ; 
+}
```
Comment (matkatmusic) [inline main.cpp]: you're multiplying twice.

- [x] FIXME

File: main.cpp [hunk 11]
```diff
@@ -61,124 +69,342 @@ New/This/Pointers/References conclusion
...
+DoubleType& DoubleType::add( const FloatType& ft )
+{
+    return add( *ft.pointerToFloatValue ); 
 }
 
+DoubleType& DoubleType::subtract( const FloatType& ft ) 
+{ 
+    *pointerToDoubleValue -= *ft.pointerToFloatValue;
+    return subtract( *ft.pointerToFloatValue ); 
+}
```
Comment (matkatmusic) [inline main.cpp]: - [x] FIXME

File: main.cpp [hunk 12]
```diff
@@ -61,124 +69,342 @@ New/This/Pointers/References conclusion
...
+
+DoubleType& DoubleType::multiply( const FloatType& ft ) 
+{ 
+    *pointerToDoubleValue *= *ft.pointerToFloatValue;
+    return multiply ( *ft.pointerToFloatValue );
+}
```
Comment (matkatmusic) [inline main.cpp]: - [x] FIXME

File: main.cpp [hunk 13]
```diff
@@ -61,124 +69,342 @@ New/This/Pointers/References conclusion
...
+DoubleType& DoubleType::add( const IntType& it )
+{
+    return add( *it.pointerToIntValue );
+}
+
+DoubleType& DoubleType::subtract( const IntType& it ) 
+{ 
+    *pointerToDoubleValue -= *it.pointerToIntValue;
+    return subtract( *it.pointerToIntValue ); 
+}
```
Comment (matkatmusic) [inline main.cpp]: - [x] FIXME

File: main.cpp [hunk 14]
```diff
@@ -61,124 +69,342 @@ New/This/Pointers/References conclusion
...
+
+DoubleType& DoubleType::multiply( const IntType& it ) 
+{ 
+    *pointerToDoubleValue *= *it.pointerToIntValue;
+    return multiply( *it.pointerToIntValue ); 
+}
```
Comment (matkatmusic) [inline main.cpp]: - [x] FIXME

**Review by matkatmusic — CHANGES_REQUESTED** [review]
FIXMEs!!!

File: main.cpp [hunk 15]
```diff
@@ -61,124 +69,335 @@ New/This/Pointers/References conclusion
...
 
-double DoubleType::divide (double x, double y) 
+DoubleType& DoubleType::divide ( const double y) 
 { 
-    if (abs(y) > 0) return (x/y);    
-    
-    std::cout << "Divide-by-zero warning  " << std::endl;
-    return (x/y);      
+    std::string warning = (abs(y) > 0) ? "" : "\x1B[31m Divide-by-zero warning! \x1B[0m";
+    std::cout << warning << std::endl;
```
Comment (matkatmusic) [inline main.cpp:185]: why not simply:
```
if( abs(y) > 0 )
    std::cout << "warning: floating point division by zero\n";
```
You're constructing a string that is potentially empty, and then printing it regardless of whether or not it is an empty string or not.

**Review by matkatmusic — APPROVED** [review]

File: main.cpp (hunk 15 above)
Comment (cristianvogel) [reply main.cpp:185]: in your suggestion, a non-zero denominator would show the warning.

**Pull request merged**